_posts/2021-08-31-modifier-type-clef-primaire-non-referencee-vers-biginit.md

```diff
@@ -0,0 +1,238 @@
+
+
+### Etape 1 : convertir les données en tâche de fond
+
+La [solution 1](https://tech.coffeemeetsbagel.com/schema-migration-from-int-to-bigint-on-a-massive-table-in-postgres-aabb835c3b84) effectue en tâche de fond la conversion des données implicite dans `ALTER TABLE foo COLUMN idBigInteger BIG INTEGER`.
+
+La [documentation](https://www.postgresql.org/docs/current/sql-altertable.html#notes) mentionne
+> Changing the type of an existing column will require the entire table and its indexes to be rewritten.
+> As an exception, when changing the type of an existing column and the old type is either binary coercible to the new type a table rewrite is not needed; but any indexes on the affected columns must still be rebuilt.
+> Table and/or index rebuilds may take a significant amount of time for a large table; and will temporarily require as much as double the disk space.
+
+Il pourrait sembler qu'aucune conversion n'est nécessaire de `INTEGER` à BIGINTEGER`, mais ce n'est pas le cas.
+```sql SELECT 1::integer::bigint```
+
+Comme l'opération native pose un verrou exclusif sur la table, il faut attendre sa réécriture complète pour y accéder.
+> An ACCESS EXCLUSIVE lock is held unless explicitly noted.
+
+### Etape 2: substituer les colonnes et créer la clef primaire
+
+Une fois la migration achevée, une fenêtre de maintenance est requise (1h dans son cas) pour :
+
+- substituer la colonne temporaire ;
+- valider les contraintes (PK et FK).
+
+### Améliorations
+
+Cette solution va dans le bon sens, mais ne satisfait par tous nos critères : il reste une fenêtre de maintenance. Nous pensons qu'il est possible de pousser la solution et d'éliminer totalement cette fenêtre.
+
+Le principe est le suivant :
+
+- la contrainte de clef primaire comporte une contrainte NOT NULL et une contrainte UNIQUE ;
+- c'est la validation de ces contraintes qui prend le plus de temps dans la fenêtre de maintenance ;
+- ces validations peuvent être effectuées en amont.
+
+Pour valider les contraintes en amont :
+
+- ajouter une contrainte NOT NULL dès la création de la colonne ;
+- créer un index UNIQUE de manière concurrente [`CREATE INDEX CONCURRENTLY`](https://www.postgresql.org/docs/current/sql-createindex.html).
+
+Lors de la création de la PK, il suffit de préciser l'index UNIQUE (PostgreSQL détecte la contrainte NOT NULL).
+`ALTER TABLE "foo" ADD CONSTRAINT "foo_pkey" PRIMARY KEY USING INDEX "foo_id_unique`
+
+## Implémentation
+
+La solution est présentée en pseudo SQL/NodeJS sur une table imaginaire, avec les liens vers les commits réels.
+
+On part de là
+
+``` sql
+CREATE TABLE foo (ID INTEGER PRIMARY KEY);
+INSERT INTO foo generate_series(1, 1e6);
+```
+
+Voilà la solution native, avec accès exclusif sur la table pendant 8 heures.
+
+```sql
+ALTER TABLE foo COLUMN idBigInteger BIG INTEGER;
```

Comment:
car j'ai l'impression que le but c'est d'avoir un truc en ready to paste

Reply:
Pour cela, il y a les références au commits
Si je mets la version SQL réelle, rien que la création du trigger prend 20 lignes